src/utils.ts

```diff
@@ -275,3 +282,53 @@ export async function calcConstrastLimits<S extends string[]>(
     })
   );
 }
+
+function getV2DataType(dtype: string) {
+  const mapping: Record<string, string> = {
+    int8: '|i1',
+    uint8: '|u1',
+    int16: '<i2',
+    uint16: '<u2',
+    int32: '<i4',
+    uint32: '<u4',
+    int64: '<i8',
+    uint64: '<u8',
+    float32: '<f4',
+    float64: '<f8',
+  };
+  if (!(dtype in mapping)) {
+    throw new Error(`Unsupported dtype ${dtype}`);
+  }
+  return mapping[dtype];
+}
+
+type Selection = (number | Omit<Slice, 'indices'> | null)[];
+
+export function createZarrArrayAdapter(arr: zarr.Array<zarr.DataType>): any {
+  return new Proxy(arr, {
+    get(target, prop) {
+      if (prop === 'getRaw') {
+        return (selection: Selection) => {
+          return get(
+            target,
+            selection.map((s) => {
+              if (typeof s === 'object' && s !== null) {
+                return slice(s.start, s.stop, s.step);
+              }
+              return s;
+            })
+          );
+        };
+      }
+      if (prop === 'getRawChunk') {
+        return (selection: number[], options: { storeOptions: RequestInit }) => {
+          return target.getChunk(selection, options.storeOptions);
+        };
+      }
+      if (prop === 'dtype') {
+        return getV2DataType(target.dtype);
+      }
+      return Reflect.get(target, prop);
+    },
+  });
+}
```

Comment:
this is the trickiest bit, which required looking at the viv source code. Basically we want to mock something like zarr.js's `ZarrArray`, but we only need a couple of methods. We could fix this by just narrowing the interface that we require in viv.

Reply:
This is gone now.